code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/load/clustering/hac/GraphHierarchicalAgglomerativeClustering.kt

```diff
@@ -0,0 +1,234 @@
+            } catch (ex: Throwable) {
+                logger.severe {
+                    """Communities merging error {$ex}:
+                        |First triple: $first
+                        |Second triple: $second
+                    """.trimMargin()
+                }
+            }
+        }
+        logger.finer { "Clustering finished" }
+        val clustering = ClusteringImpl(communities.map { it.entities.toSet() })
+        val clustersGraph = buildClustersGraph()
+        return ClusteringResult(clustering, clustersGraph)
+    }
+
+    private fun buildClustersGraph(): ClustersGraph<V> {
+        val clustrersGraph = SimpleWeightedGraph<Cluster<V>, DefaultWeightedEdge>(DefaultWeightedEdge::class.java)
+        triples.values.forEach { triple ->
+            val firstVertex = triple.first.entities.toSet()
+            val secondVertex = triple.second.entities.toSet()
+            clustrersGraph.addVertex(firstVertex)
+            clustrersGraph.addVertex(secondVertex)
+            clustrersGraph.addEdge(firstVertex, secondVertex)
+            val edge = clustrersGraph.getEdge(firstVertex, secondVertex)
+            clustrersGraph.setEdgeWeight(edge, triple.distance)
+        }
+        return clustrersGraph
+    }
+
+    private fun mergeCommunities(first: Community, second: Community) {
+        val merged: MutableList<V>
+        if (first.entities.size < second.entities.size) {
+            merged = second.entities
+            merged.addAll(first.entities)
+        } else {
+            merged = first.entities
+            merged.addAll(second.entities)
+        }
+        val newCommunity = Community(merged)
+        communities.remove(first)
+        communities.remove(second)
+        for (community in communities) {
+            val fromFirstId = getTripleId(first, community)
+            val fromSecondId = getTripleId(second, community)
+            val fromFirst = triples[fromFirstId]
+            val fromSecond = triples[fromSecondId]
+            val newDistance = getDistance(fromFirst).coerceAtLeast(getDistance(fromSecond))
+            invalidateTriple(fromFirst)
+            invalidateTriple(fromSecond)
+            insertTriple(newDistance, newCommunity, community)
+        }
+        communities.add(newCommunity)
+    }
+
+    private fun getDistance(triple: Triple?): Double = triple?.distance ?: Double.POSITIVE_INFINITY
+
+    private fun getTripleId(first: Community, second: Community): Long = if (second.id > first.id) {
+        getTripleId(second, first)
+    } else {
+        first.id * ID_FACTOR + second.id
+    }
+
+    private fun insertTriple(triple: Triple) {
+        triples[getTripleId(triple.first, triple.second)] = triple
+        if (triple.distance < distanceLimit) {
+            heap.add(triple)
+        }
+    }
+
+    private fun insertTriple(distance: Double, first: Community, second: Community) {
+        val triple = createTriple(distance, first, second)
```

Comment:
`createTriple(distance, first, second).also {
            insertTriple(it)
        }`